hw06 Annotations

freegate78 opened the pull request:
start main from hw06_main class

freegate78 committed:
hw03 ArrayListDemo

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/Before.java:
```diff
@@ -0,0 +1,8 @@
+package com.akproductions;
```
Хорошо бы разнести классы по пакетам.
Подумайте сами по каким.

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/hw06_main.java:
```diff
@@ -0,0 +1,124 @@
+    private MessageTemplateProvider provider;
+    private MessageBuilder messageBuilder;
+
+    public static void main(String[] args) throws Exception {
```
В `main()` не должно быть самого тестового фрейворка.
Он должен быть инкапсулирован в каком-то классе. 
Здесь должен быть просто его вызов:
```java
TestResult results = TestFramework.run("com.akproductions.MessageBuilderImpl")
```
И лучше именно так - получить результаты в класс `TestResult`, а потом уже выводить их как надо и куда надо.

Также у вас main() очень большой. Подумайте какие методы в нем можно выделить.

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/hw06_main.java:
```diff
@@ -0,0 +1,124 @@
+
+    public static void main(String[] args) throws Exception {
+        //входной параметр - имя класса
+        String ClassToTest = "com.akproductions.MessageBuilderImpl";
```
```suggestion
        String classToTest = "com.akproductions.MessageBuilderImpl";
```

Переменные принято писать в camelCase

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/hw06_main.java:
```diff
@@ -0,0 +1,124 @@
+        //входной параметр - имя класса
+        String ClassToTest = "com.akproductions.MessageBuilderImpl";
+        Class clazz = Class.forName(ClassToTest);
+        //Class<MessageBuilderImpl> clazz = MessageBuilderImpl.class;
```
Закоментированный код не надо комитить.
Не понятно, зачем он нужен и почему закоментирован.

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/hw06_main.java:
```diff
@@ -0,0 +1,124 @@
+
+        //System.out.println("--- annotations:");
+        //методы читсто в нашем классе - чтобы следить за аннотациями
+        Method[] methodsAll = clazz.getDeclaredMethods();
+        Method[] methodsBefore = new Method[100] ;
+        Method[] methodsAfter = new Method[100] ;
+        Method[] methodsTest = new Method[100] ;
```
Как-то странно ограничивать длину в 100 элементов.
Подумайте, что можно использовать вместо массива, чтобы не было ограничений на количество элементов.

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/hw06_main.java:
```diff
@@ -0,0 +1,124 @@
+            Annotation[] annotations = method.getDeclaredAnnotations();
+            for (Annotation annotation : annotations) {
+                switch (annotation.toString()){
+                    case  ("@com.akproductions.Before()"):
```
Лучше сравнивать не строки, а сами классы аннотаций.
```java
annotation.equals(Before.class)
```

А почему - я спрошу у вас.

Да, от case придется уйти, использовать if. Но это не проблема.

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/SimpleAnnotation.java:
```diff
@@ -0,0 +1,9 @@
+
+
+@Retention(RetentionPolicy.RUNTIME)
+public @interface SimpleAnnotation {
```
Вроде не используется.
Зачем нужна эта аннотация?
Лучше убрать.

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/hw06_main.java:
```diff
@@ -0,0 +1,124 @@
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+public class hw06_main {
```
```suggestion
public class HW06Main {
```

Принято классы именовать с большой буквы.

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/hw06_main.java:
```diff
@@ -0,0 +1,124 @@
+                        //System.out.println(Arrays.toString(constructors));
+
+                        //System.out.println("--- creating new object:");
+                        MessageTemplateProvider provider = new MessageTemplateProvider();
+                        Constructor<MessageBuilderImpl> constructor = clazz.getConstructor(MessageTemplateProvider.class);
+                        MessageBuilderImpl object = constructor.newInstance(provider);
```
Здесь не должно быть никаких упоминаний MessageTemplateProvider, MessageBuilderImpl и др классов, которые вы тестируете.
У вас имя класса с тестами уже есть в  `ClassToTest` и экземпляр этого класса и надо здесь создать.

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/hw06_main.java:
```diff
@@ -0,0 +1,124 @@
+                        System.out.println("[DEBUG] --- invoke test method  " + methodsTest[i].getName());
+                        methodsTest[i].setAccessible(true);
+
+                        //2 вида запуска - с аргументом и без проверяем
```
Давайте пока не будем усложнять и делать тесты с аргументами.
Просто уберите это. Оставьте простые тесты без аргументов.

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/hw06_main.java:
```diff
@@ -0,0 +1,124 @@
+                        try {
+                            result = methodsTest[i].invoke(object);
+                            System.out.println("[DEBUG] --- invoked test method  " + methodsTest[i].getName() + " without arguments");
+                        } catch (IllegalArgumentException e1) {
+                                result = methodsTest[i].invoke(object, 0);
+                                System.out.println("[DEBUG] --- invoked test method  " + methodsTest[i].getName() + " with arguments");
+                            }
+                        catch (InvocationTargetException e1) {
+                            System.out.println("[DEBUG] --- invoked test method  " + methodsTest[i].getName() + " without arguments has target exception, reason: " + e1.getTargetException());
+                            throw e1;
+                        }
+                        catch (Exception e1) {
+                            System.out.println("[DEBUG] --- invoked test method  " + methodsTest[i].getName() + " without arguments failed, reason: " + e1.getLocalizedMessage());
+                            throw e1;
+                        }
```
Смысла обрабатывать каждое исключение нет.
Достаточно
```java
catch (Exception e1)
```

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/hw06_main.java:
```diff
@@ -0,0 +1,124 @@
+                        }
+                        System.out.println("[DEBUG] --- test method result is: " + result);
+
+                        for (int k = 0; k < cAfter; k++) {
+                            System.out.println("[DEBUG] --- invoke after method  " + methodsAfter[k].getName());
+                            methodsAfter[k].setAccessible(true);
+                            methodsAfter[k].invoke(object);
+                        }
```
after-методы должны выполняться всегда, вне зависимости от того была ли ошибка в тесте или нет.
Т.к. здесь может быть закрытие какие-то ресурсов.
Подумайте, как это можно сделать.

freegate78 committed:
hw03 ArrayListDemo

spvgl commented on HW06-Annotation/src/main/java/com/testframework/Hw06_main.java:
```diff
@@ -0,0 +1,9 @@
+public class Hw06_main {
+
+    public static void main(String[] args) throws Exception {
+        String ClassToTest = "com.akproductions.MessageBuilderImpl";
```
```suggestion
        String classToTest = "com.akproductions.MessageBuilderImpl";
```

spvgl commented on HW06-Annotation/src/main/java/com/testframework/TestFramework.java:
```diff
@@ -0,0 +1,98 @@
+import java.util.concurrent.atomic.AtomicInteger;
+
+public class TestFramework {
+    Class clazz;
```
```suggestion
private Class clazz;
```

Поля должны быть приватными.
Инкапсуляция.
Чтобы никто снаружи не мог напрямую изменить состояние класса.

spvgl commented on HW06-Annotation/src/main/java/com/testframework/TestFramework.java:
```diff
@@ -0,0 +1,98 @@
+        clazz = Class.forName(classToTest);
+    }
+
+    public String doTest () throws Exception
```
Результаты лучше вернуть не строкой, а в виде класса, в котором будет информация о количестве тестов, количестве удачных, неудачных и тп.
Эту информацию вы уже сможете вывести в любом формате.

Еще можно подумать с точки зрения тестирования самого вашего тестового фреймворка.
При возврате строкой проверить правильность работы вашего фреймворка будет сложнее.

spvgl commented on HW06-Annotation/src/main/java/com/testframework/TestFramework.java:
```diff
@@ -0,0 +1,98 @@
+        var methodsAfter = new ArrayList<Method>(10);
+        var methodsTest = new ArrayList<Method>(10);
+        AtomicInteger testSuccess= new AtomicInteger();
+        AtomicInteger testError= new AtomicInteger();
```
Почему решили применить AtomicInteger, а не простой int?

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/MessageBuilderImpl.java:
```diff
@@ -0,0 +1,90 @@
+package com.akproductions;
+
+public class MessageBuilderImpl implements MessageBuilder {
```
У вас это и тест и класс, который тестируете.
Лучше разделить на два класса - `MessageBuilderImpl` и `MessageBuilderImplTest`

spvgl commented on HW06-Annotation/src/main/java/com/testframework/TestFramework.java:
```diff
@@ -0,0 +1,98 @@
+                            System.out.println("[ERROR] --- error in calling before method  " + methodBefore.getName());
+                            e.printStackTrace(System.out);
+                        }
+                    });
```
Можно выделить в метод:
```suggestion
                    runMethods(methodsBefore, object);
```

Для After-методов то же самое.

freegate78 committed:
propagade to custom packages
 app - all business logic
 testframework - all with running tests
 tests - test copy of business logic for unit tests

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/testframework/TestFramework.java:
```diff
@@ -0,0 +1,83 @@
+        return testResults;
+    }
+
+    private void invokeMethods(ArrayList<Method> methods,Object object, String typeOfMethod)
```
```suggestion
    private void invokeMethods(List<Method> methods,Object object, String typeOfMethod)
```

Лучше указывать интерфейс, чтобы можно было передать любой список

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/testframework/TestFramework.java:
```diff
@@ -0,0 +1,83 @@
+        methodsTest.forEach(c ->   System.out.println(c.getName()));
+
+        //make tests
+        methodsTest.forEach(methodTest ->{
+            try {
+                var object = clazz.getConstructor().newInstance();
+                invokeMethods(methodsBefore,object,"Before");
+                methodTest.setAccessible(true);
+                try {
+                    methodTest.invoke(object);
+                    testSuccess.getAndIncrement();
+                    System.out.println("[INFO] --- test "+methodTest.getName()+" completed successfully");
+                } catch (Exception e) {
+                    testError.getAndIncrement();
+                    System.out.println("[ERROR] --- error in calling test method  " + methodTest.getName()+ ". reason: " + e.getLocalizedMessage());
+                } finally {invokeMethods(methodsAfter,object,"After");}
+            } catch (Exception e) {
+                testError.getAndIncrement();
+                System.out.println("[FATAL ERROR] --- running all test scenario for method " + methodTest.getName() + " failed, reason is: " + e.getLocalizedMessage());
+            }
+        });
```
Когда хочется написать подобный комментарий "make tests" сразу напрашивается  выделить весь этот кусок кода в метод `makeTests()`. И комментарий тогда не будет нужен

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/testframework/TestFramework.java:
```diff
@@ -0,0 +1,83 @@
+        testResults= new HashMap<String,Integer> ();
+    }
+
+    public HashMap<String,Integer> doTest () throws Exception
```
Давайте еще улучшим и заменим HashMap на класс.
Не бойтесь добавлять новые классы.
У нас ООП. 

Например, посмотрите на такую сигнатуру метода:
```java
public TestResult doTest()
```
Сразу видно по сигнатуре, что возвращается результат тестирования.

А в случае:
```java
public HashMap<String,Integer> doTest()
```
Какая-то мапа со строкой в ключе и числом в значении.
Да, можно посмотреть, изучить код и понять, что там возвращается, но это трата лишних сил, времени и тп 
И с мапой со строковыми ключами можно сделать ошибки в них, с классом у вас будет проверка во время компиляции.

Еще момент, в таких случаях лучше определять через интерфейс:
```suggestion
    public Map<String,Integer> doTest()
```

freegate78 committed:
propagade to custom packages
 app - all business logic
 testframework - all with running tests
 tests - test copy of business logic for unit tests

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/testframework/TestResults.java:
```diff
@@ -0,0 +1,14 @@
+
+import java.util.HashMap;
+
+public class TestResults<String, Integer> extends HashMap
```
С таким классом, по сравнению с ранее использованным HashMap, у вас ничего по сути не изменилось.
Когда вы его используете:
```java
testResults.put(methodTest.getName(), 1);
```
не понятно, что такое 0 и 1 в качестве значения.
Можно догадаться, что это успешное и не успешное выполнение теста, но чтобы это понять надо изучить код, где это используется.
Здесь как минимум надо boolean вместо 0 и 1.

А вообще TestResult я имел ввиду сделать примерно таким:
```java
public class TestResults  {
    private int successCount;
    private int failedCount;
    private int totalCount;
    // ... конструктор
    // ... геттеры
}
```
Т.к.  по заданию нужна только сводка.
Можно расширить его как вы сделали - хранить результаты каждого теста.
```java
public class TestResults {
    private int successCount;
    private int failedCount;
    private int totalCount;

    private List<TestResult> results;

    // ... конструктор
    // ... геттеры
    
    static class TestResult {
        private Method method;
        private boolean isSuccess;
        private Exception exception; // или текст private String errorDescriptoin;
        // ... конструктор
        // ... геттеры
    }
}
```

Глядя на такой класс сразу видно, что в нем хранится.
Глядя на:
```java
public class TestResults<String, Integer> extends HashMap {
```
не понятно, что и в каком виде хранится в таком TestResults<String, Integer>.

Еще здесь у вас форматирование кода странное. Нажмите Ctrl + Alt + L в IDEA

spvgl commented on HW06-Annotation/src/main/java/com/akproductions/testframework/TestFramework.java:
```diff
@@ -0,0 +1,85 @@
+        return makeTests(methodsTest, methodsBefore, methodsAfter);
+    }
+
+    private Map<String, Integer> makeTests(ArrayList<Method> methodsTest,ArrayList<Method> methodsBefore, ArrayList<Method> methodsAfter) throws Exception
```
```suggestion
    private Map<String, Integer> makeTests(List<Method> methodsTest, List<Method> methodsBefore, List<Method> methodsAfter) throws Exception
```

Интерфейсы вместо реализаций

freegate78 committed:
1) преобразовал класс результатов по рекомендации - не HashMap, а конкретные метрики тестов, а также методы для их установки и изменения
2) исправил параметры методов на интерфейс map <Method>

freegate78 commented:
Андрей Кузьмин05.05.2020 в 22:32
https://github.com/freegate78/java-2020-03/pull/3
06.05.2020 в 07:16 Павел Стрекалов назначил проверяющего
06.05.2020 в 07:16 Павел Стрекалов взял ДЗ на проверку
Павел Стрекалов06.05.2020 в 09:19
Здравствуйте, Андрей.

Чувствуется вы на C/C++ много программировали.
В java стиль немного другой - см. замечания в гитхабе.
06.05.2020 в 09:19 Павел Стрекалов вернул ДЗ на доработку
Андрей Кузьмин07.05.2020 в 14:08
добавил коммит в тот же ПР
https://github.com/freegate78/java-2020-03/pull/3/commits/ef7e9f6eac9476b1dd28f3ae7144d7996f534993
сделал в отдельном package класс для тестов, и его простой вызов в мейн
изменил массивы на коллекции для методов
сделал отбработку after через finally
убрал c-like стиль программирования, перевел все циклы на методы или лямбды

из непонятного - почему то исключение при вызове теста
[ERROR] --- error in calling test method buildMessage3
пробрасывается в stderr, хотя я везде поставил обработчики. пока не смог понять почему.
Caused by: java.lang.Exception: [ERROR] class com.akproductions.TemplateNotFoundException

лог тестов
> Task :HW06-Annotation:Hw06_main.main()
[INFO] --- Before methods:
doBefore
[INFO] --- After methods:
doAfter
[INFO] --- Test methods:
doTest
buildMessage1
buildMessage3
buildMessage2
[INFO] --- run test for doTest
[DEBUG] --- invoke before method doBefore
Before called!!! var1=0, var2=0
[DEBUG] --- invoke test method doTest
Test called!!! var1=0, var2=0
[INFO] --- test runned successfully
[DEBUG] --- invoke before method doAfter
After called!!! var1=0, var2=0
[INFO] --- run test for buildMessage1
[DEBUG] --- invoke before method doBefore
Before called!!! var1=0, var2=0
[DEBUG] --- invoke test method buildMessage1
[INFO] --- test runned successfully
[DEBUG] --- invoke before method doAfter
After called!!! var1=0, var2=0
[INFO] --- run test for buildMessage3
[DEBUG] --- invoke before method doBefore
Before called!!! var1=0, var2=0
[DEBUG] --- invoke test method buildMessage3
[ERROR] --- error in calling test method buildMessage3
java.lang.reflect.InvocationTargetException
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.base/java.lang.reflect.Method.invoke(Method.java:567)
at com.testframework.TestFramework.lambda$doTest$5(TestFramework.java:65)
at java.base/java.util.ArrayList.forEach(ArrayList.java:1507)
at com.testframework.TestFramework.doTest(TestFramework.java:43)
at com.testframework.Hw06_main.main(Hw06_main.java:7)
Caused by: java.lang.Exception: [ERROR] class com.akproductions.TemplateNotFoundException
at com.akproductions.MessageBuilderImpl.buildMessage3(MessageBuilderImpl.java:87)
... 8 more
Caused by: java.lang.Exception: [ERROR] class com.akproductions.TemplateNotFoundException

[DEBUG] --- invoke before method doAfter
After called!!! var1=0, var2=0
[INFO] --- run test for buildMessage2
[DEBUG] --- invoke before method doBefore
Before called!!! var1=0, var2=0
[DEBUG] --- invoke test method buildMessage2
[INFO] --- test runned successfully
[DEBUG] --- invoke before method doAfter
After called!!! var1=0, var2=0
[INFO] --- total tests runned: 4, successfully runned: 3, not runned because of an error: 1

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.3/userguide/command_line_interface.html#sec:command_line_warnings
07.05.2020 в 18:51 Павел Стрекалов взял ДЗ на проверку
Павел Стрекалов08.05.2020 в 09:08
>> сделал в отдельном package класс для тестов, и его простой вызов в мейн

См. вложение, как предлагаю сделать пакеты.
Если есть вопросы, почему так, то задавайте.
Павел Стрекалов08.05.2020 в 09:09
packages-188-774289.png
9 КБ
Павел Стрекалов08.05.2020 в 09:09
>> перевел все циклы на методы или лямбды

Много лямбд это тоже плохо.
Часто они применяются не к месту и ухудшают читаемость кода.

Например, здесь
methodsTest.forEach(methodTest ->
{
...
}

Я бы так не писал. Очень большой кусов кода в лямбде.
Когда его дочитаешь до конца, листая экран, уже забывается, что он внутри лямбды.
Его надо выделить в метод, может не весь, может часть, подумайте.

>> из непонятного - почему то исключение при вызове теста
>> [ERROR] --- error in calling test method buildMessage3
>> пробрасывается в stderr, хотя я везде поставил обработчики. пока не смог понять почему.

Вы имеете ввиду вот это?
e.printStackTrace(System.out);
Или что?

Чем вас stderr смущает? Лучше не переделывать эту логику, что ошибки в stderr.
Все привыкли, что ошибки там и могут быть неудобства, если кто-то запустит ваше приложение и будет ожидать ошибки в stderr, а их там не будет.
Это я не касательно этого ДЗ, а в целом.

Лучше здесь стек не выводить, а вывести только то, исключение, которое возникло в тесте.
Павел Стрекалов08.05.2020 в 09:10
Еще см. комментарии в гитхабе
08.05.2020 в 09:10 Павел Стрекалов вернул ДЗ на доработку
Андрей Кузьмин08.05.2020 в 10:53
https://github.com/freegate78/java-2020-03/pull/3/commits/099fd9a12c50b4b973844ce769c1cc3417f612d1
08.05.2020 в 15:56 Павел Стрекалов взял ДЗ на проверку
Павел Стрекалов08.05.2020 в 16:25
Молодцы.
Видите, после выделения метода стало гораздо читаемее.
Эта часть (methodsTest.forEach) даже на один экран стала влазить :)

По поводу AtomicInteger вы не пояснили. Зачем он нужен?

Мы на финишной прямой.
Еще несколько замечаний в гитхабе.
08.05.2020 в 16:25 Павел Стрекалов вернул ДЗ на доработку
Андрей Кузьмин11.05.2020 в 23:17
про атомик - там же были try блоки, идеа не дает использовать обычный int, и не компилит, видимо атомик - спец тип, который везде виден и меняется. я его в итоге убрал, т.к. вывел запуск тестов и сбор результатов в отдельный метод, где сразу наполняю коллекцию результатами. на комменты дал фидбек https://github.com/freegate78/java-2020-03/pull/3/commits/099fd9a12c50b4b973844ce769c1cc3417f612d1, сделал новый коммит https://github.com/freegate78/java-2020-03/pull/3/commits/d57761b017f208dabf48662d1ec8f4b0230b27e4
Павел Стрекалов08:37 (ред.)
>> про атомик - там же были try блоки, идеа не дает использовать обычный int,

Атомики и try никак не связаны.
Скорее всего, ваша ошибка была связана со Stream API.
Внутри лямбды нельзя модифицировать пременные.
Они должны быть final или эффективно final.

И все-такие я бы вместо этой лямбды:
methodsTest.forEach(methodTest -> ...
сделал бы простой цикл for each.
Код был бы проще. Меньше было бы путаницы и непонимания.

Лямбды хорошо использовать, когда они помогают превратить анонимный класс из десятка строк в одну строку.
Но когда сама лямбда на десяток строк - это уже выглядит хуже.
Хотя, стоит отметить, что по сравнению с первоначальным вариантом стало значительно лучше.

Почему я так много об этом пишу.
Потому, что чаще код читают, чем пишут. И он должен читаться просто. В идеале так же просто, как вы читаете книгу - строка за строкой без каких-то WTF на каждой строке. Если для понимания того, что происходит в какой-то строке кода, надо изучить еще десяток классов/файлов, то это трата времени и вероятность совершить ошибку. Но это в идеале, к которому надо стремиться. В реальности, конечно, бывает всякое :)

>> на комменты дал фидбек https://github.com/freegate78/java-2020-03/pull/3/commits/099fd9a12c50b4b973844ce769c1cc3417f612d1

Не нашел вашего фидбека

ДЗ принимаю. Поработали хорошо. Если что-то поправите, то можете присылать сюда - проверю.
В гитхабе еще оставил комментарий.
08:40 Павел Стрекалов принял ДЗ
OTUS08:40
Уважаемый студент! Как вам проверка домашних заданий в Otus?
Расскажите в опросе.
Андрей Кузьмин11:17
1) преобразовал класс результатов по рекомендации - не HashMap, а конкретные метрики тестов, а также методы для их установки и изменения
2) исправил параметры методов на интерфейс map <Method>

https://github.com/freegate78/java-2020-03/pull/3/commits/b2d9185c6c275789507a335eeab41159ccead45f
Павел Стрекалов17:20
Все хорошо. Исправления приняты.
Андрей Кузьмин17:31
ok тогда мерджу и к след перехожу
